File: LC/lc198/lc198.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var rob = function(nums) {
  var res = [0, 0];
  for (i = 0; i < nums.length; i++){
      res = [Math.max(res[1], res[0]), res[0] + nums[i]];
  }
  return Math.max(res[0], res[1]);
};

File: LC/lc198/lc198.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        if not nums:
            return 0
        if len(nums) == 1:
            return nums[0]
        dp = {0:(nums[0], True)}
        if nums[1] > nums[0]:
            dp[1] = (nums[1], True)
        else:
            dp[1] = (nums[0], False)
        k = len(nums) - 1
        for i in range(2, k+1):
            if dp[i-1][1] == False:
                dp[i] = (dp[i-1][0] + nums[i], True)
            else:
                left = dp[i-2][0] + nums[i]
                right = dp[i-1][0]
                if left >= right:
                    dp[i] = (left, True)
                else:
                    dp[i] = (right, False)
        return dp[k][0]